react/features/videosipgw/middleware.js

```diff
@@ -0,0 +1,182 @@
+/* @flow */
+
+import Logger from 'jitsi-meet-logger';
+import { CONFERENCE_WILL_JOIN } from '../base/conference';
+import {
+    SIP_GW_AVAILABILITY_CHANGED,
+    SIP_GW_INVITE_ROOMS
+} from './actionTypes';
+import {
+    JitsiConferenceEvents,
+    JitsiSIPVideoGWStatus
+} from '../base/lib-jitsi-meet';
+import { MiddlewareRegistry } from '../base/redux';
+import {
+    Notification,
+    showErrorNotification,
+    showNotification,
+    showWarningNotification
+} from '../notifications';
+
+const logger = Logger.getLogger(__filename);
+
+/**
+ * Middleware that captures conference video sip gw events and stores
+ * the global sip gw availability in redux or show appropriate notification
+ * for sip gw sessions.
+ * Captures invitation actions that create sip gw sessions or display
+ * appropriate error/warning notifications.
+ *
+ * @param {Store} store - The redux store.
+ * @returns {Function}
+ */
+MiddlewareRegistry.register(({ dispatch, getState }) => next => action => {
+    const result = next(action);
+
+    switch (action.type) {
+    case CONFERENCE_WILL_JOIN: {
+        const conference = getState()['features/base/conference'].joining;
+
+        conference.on(
+            JitsiConferenceEvents.VIDEO_SIP_GW_AVAILABILITY_CHANGED,
+            (...args) => dispatch(_availabilityChanged(...args)));
+        conference.on(
+            JitsiConferenceEvents.VIDEO_SIP_GW_SESSION_STATE_CHANGED,
+            event => {
+                const toDispatch = _sessionStateChanged(event);
+
+                // sessionStateChanged can decide there is nothing to dispatch
+                if (toDispatch) {
+                    dispatch(toDispatch);
+                }
+            });
```

Comment:
Could you tell me how the timing of the join flow works? connection.initJitsiConference is called, which the browser takes over (async), but in the same stack as the call to  connection.initJitsiConference the action CONFERENCE_WILL_JOIN is dispatched? I want to make sure it is guaranteed these conference listeners are set before the video sip gw events can be fired.

Reply:
So after we call initJitsiConference we fire the CONFERENCE_WILL_JOIN action, and after the call for initJitsiConference returns the conference object we call on it join.
Basically I was added the same that mobile is currently doing:
https://github.com/jitsi/jitsi-meet/blob/b91d588f9ceb6f4462576a3517067dc9e56610f8/react/features/base/conference/actions.js#L304